### ISSUE TITLE
[Bug]: window function: frame range's poor performance.

### ISSUE THREAD

**Issue by Ariznawlll** [matrixorigin/matrixone#13008]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```Markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93):  0015299e158756706f2f953348229ca889a1f649
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

<img width="853" alt="image" src="https://github.com/matrixorigin/matrixone/assets/108530700/36724345-8e9a-48e8-ae3c-1c2a00955f9c">

### Expected Behavior

<img width="697" alt="image" src="https://github.com/matrixorigin/matrixone/assets/108530700/f6ac57bc-b9b2-4dfb-8bd1-1b339dad54b3">

### Steps to Reproduce

```Markdown
drop table if exists td;
create table td(d int);

insert into td(d) values (10),(1),(2),(3),(4),(5),(6),(7),(8),(9);
insert into td(d) select d+10 from td;
insert into td(d) select d+20 from td;
insert into td(d) select d+30 from td;
insert into td(d) select d+40 from td;
insert into td(d) select d+50 from td;
insert into td(d) select d+60 from td;
insert into td(d) select d+70 from td;
insert into td(d) select d+80 from td;
insert into td(d) select d+90 from td;

insert into td(d) select d+100 from td;
insert into td(d) select d+200 from td;
insert into td(d) select d+300 from td;
insert into td(d) select d+400 from td;
insert into td(d) select d+500 from td;
insert into td(d) select d+600 from td;
insert into td(d) select d+700 from td;
insert into td(d) select d+800 from td;
insert into td(d) select d+900 from td;
insert into td(d) select d+1000 from td;
insert into td(d) select d+2000 from td;
insert into td(d) select d+3000 from td;
select count(*) from td;

select avg(d) over (order by d range between 2 preceding and 2 following) from td limit 10;
```

**Comment by ouyuanning** [comment]
@iamlinjunhong @Ariznawlll 
在1.2有专门建了优化类的issue处理（https://github.com/matrixorigin/matrixone/issues/12961），这个bug类的issue建议先关闭吧。这个更像是一个需求

**Comment by matrix-meow** [comment]
Hello @Ariznawlll. The bug issue in the BVT test code has not been removed,issues automatically open.

**Comment by iamlinjunhong** [comment]
对 #13008 的有限边界 `RANGE AVG` 做了第一轮优化，下面是 MatrixOne 的环境与 before/after 数据。

### 测试环境

- CPU：Apple M4，10 logical CPUs（`GOMAXPROCS=10`）
- OS：macOS 15.5（24F74），arm64
- MatrixOne 基线：`main@33af6ae3c0c23e4f277137460577ea64678f7ecf`
- Go：`go1.26.4 darwin/arm64`
- 部署：本地单节点，`etc/launch/launch.toml`
- 数据：完全使用 issue 中的 fixture，共 20,971,520 行
- 计时：`/usr/bin/time -p` 的 real time；数据已预先加载，SQL 客户端连接开销计入

原始查询：

```sql
select avg(d) over (
  order by d
  range between 2 preceding and 2 following
) from td limit 10;
```

为避免 `LIMIT 10` 掩盖结果消费成本，另测了完整结果消费：

```sql
select sum(x)
from (
  select avg(d) over (
    order by d
    range between 2 preceding and 2 following
  ) x
  from td
) s;
```

### MatrixOne 端到端结果

| 版本 / 场景 | real time |
| --- | ---: |
| issue 原始报告（`0015299`） | 约 5 小时 |
| 优化后，原始 `LIMIT 10` 查询 | 1.13–1.68 s，中位数 1.54 s |
| 优化后，完整消费 20,971,520 行结果 | 2.38–2.58 s |

原始报告与当前 main 跨越较大，因此没有直接用“5 小时 → 1.54 秒”计算加速比。当前 main 的旧逐行 frame evaluator 跑完整数据仍属于小时级；为得到可重复、同版本的 before/after，下面使用只开关新 fast path 的受控 benchmark。

### 同版本 evaluator before/after

永久 benchmark：`BenchmarkWindowBoundedRangeAvg`。80,000 行已排序 `int32` 数据，frame 同样为 `RANGE 2 PRECEDING / 2 FOLLOWING`；timer 只覆盖 window evaluator，不包含数据构造和排序。每项 3 次，表中取中位数。

| 每个 order key 的 peer 行数 | 优化前（旧逐行 frame） | 优化后 | 加速 |
| ---: | ---: | ---: | ---: |
| 1 | 70.73 ms/op | 12.99 ms/op | 5.4x |
| 16 | 980.71 ms/op | 4.57 ms/op | 215x |
| 64 | 3.886 s/op | 4.13 ms/op | 940x |

issue 数据本身只有 10,960 个 distinct order key，最大 peer 为 3,848 行，按行加权的平均 peer 为 2,757 行。旧实现估算需要约 2891 亿次 aggregate `Fill`；滑动窗口只需约 4194 万次边界 add/remove，aggregate 工作量理论上减少约 6,893 倍。这也是 peer 越大，加速越明显的原因。

### 优化内容

- 对单 order key、有限边界的 `RANGE` frame 使用单调滑动聚合；
- 同一 peer group 只做一次边界搜索和状态更新，结果复用给所有 peer 行；
- integer / unsigned / bit / Decimal64 的 `AVG` 支持精确 add/remove；
- 单 group 滑动更新绕过通用 `BatchFill` 的 256-slot hash 初始化；
- float、DISTINCT、TIMESTAMP/DST fold 等不安全场景继续走原 evaluator，避免语义变化。

### 正确性与回归

- 原始查询前 10 行结果：`2, 2.5, 3, 4, 5, 6, 7, 8, 9, 10`
- `./pkg/sql/colexec/aggexec`、`./pkg/sql/colexec/window` 全量 UT 通过
- 上述 package 的 `go vet` 通过
- `make build` 通过
- 最终二进制执行 `window.sql` BVT：1106/1106 passed，0 failed，0 ignored（25.077 s）

**Comment by Ariznawlll** [comment]
### 测试版本

- 分支：`main`
- commit：`cd04bb4c1af5bc595e2147dc645dfa754f4c395b`
- 部署方式：TKE 分布式，`3 CN / 1 DN / 3 Log / 2 Proxy`
- 测试日期：2026-09-09

### 原问题验证

按 issue 的窗口函数场景构造 `20,971,520` 行数据（`d` 的范围为 `1..10960`），执行：

```sql
SELECT AVG(d) OVER (
  ORDER BY d RANGE BETWEEN 2 PRECEDING AND 2 FOLLOWING
) FROM td LIMIT 10;
```

连续执行 3 次，前三行均稳定为 `2`、`2.5`、`3`，第十行为 `10`；查询均正常返回，未出现历史上的异常性能/执行问题。

### Happy Path 覆盖

1. 原始 `RANGE BETWEEN 2 PRECEDING AND 2 FOLLOWING` 窗口聚合连续 3 次结果一致。
2. MOTR 的确定性缩小语义用例覆盖乱序写入的 `1..10`：窗口结果依次为 `2.0000`、`2.5000`、`3.0000`、`4.0000`、`5.0000`、`6.0000`、`7.0000`、`8.0000`、`8.5000`、`9.0000`；聚合校验和为 `55.0000`。

### Unhappy Path 覆盖

1. 反向 frame `RANGE BETWEEN 2 FOLLOWING AND 2 PRECEDING` 被稳定拒绝为 `ERROR 1064`；随后行数、最小值、最大值仍为 `10 / 1 / 10`，确认失败不会破坏表数据。

### 白盒验证

不适用：本 issue 是公开 SQL 窗口语义及大表执行场景；覆盖以 TKE 原始规模验收和可重复的 SQL 回归用例为准。

### 回归测试

- 已加入 MOTR PR [#152](https://github.com/matrixorigin/motr/pull/152)。
- 用例：`suites/14_issue_regression/t/issue_13008_large_range_window.test`。
- 用例单独连续运行 3 次均通过（`30` statements，`30` ok）；MOTR 负责固定语义与非法 frame 防回归，原始 `20,971,520` 行规模已在 TKE 验收，保留为专项规模覆盖。

### 测试证据

- 修复 PR [matrixone#27940](https://github.com/matrixorigin/matrixone/pull/27940) 的 merge commit `522285b47ac7c866333e47a2a680b1b38b052677` 已包含在本次环境版本中。
- MOTR 回归提交：[0028f09b](https://github.com/Ariznawlll/motr/commit/0028f09b6be54cd936530bf9f02645372e5bb57d)。

### 测试结论

测试通过。
